### ISSUE TITLE
Handling Braintree hosted fields in a way that stops us from handling credit card information directly

### ISSUE THREAD

**Issue by CraigRobertWhite** [bigcommerce/checkout-sdk-js#976]
We are currently using the [React example](https://github.com/bigcommerce/checkout-sdk-js-example) for implementing the Checkout SDK with the Cornerstone theme.

Upon reading the README for this repo, we saw that to submit an order, we should provide the `service.submitOrder` function with a payment. ([source](https://github.com/bigcommerce/checkout-sdk-js#submit-order))

This way of creating a payment requires us to handle the value of inputs directly.

Two questions:
1. Are storing the values of the customer's payment method when passing it to the `submitOrder()` function PCI compliant?
2. Is there a way to use Braintree's [Hosted Fields](https://developers.braintreepayments.com/reference/client-reference/javascript/v2/hosted-fields) in a way that does not require us to handle customer payment method information directly?
   - Is there a way to have the Hosted Fields generate a nonce token, then process the order with that nonce token?

**Comment by joeldentici** [comment]
You can use the BigCommerce hosted fields implementation that was released back in April:

Here is the [source code in checkout-js](https://github.com/bigcommerce/checkout-js/blob/bfb06fec56db840713be8773ed2a35826bce6612/src/app/payment/hostedCreditCard/withHostedCreditCardFieldset.tsx) that returns the options to initialize the payment method.

Below is where that is called from to pass those options to the checkout sdk's `CheckoutService['initializePayment']`:

```ts
    const initializeHostedCreditCardPayment: CreditCardPaymentMethodProps['initializePayment'] = useCallback(async (options, selectedInstrument) => {
        return initializePayment({
            ...options,
            creditCard: getHostedFormOptions && {
                form: await getHostedFormOptions(selectedInstrument),
            },
        });
    }, [
        getHostedFormOptions,
        initializePayment,
    ]);
```

After that, you should be able to submit the order with:

```ts
submitOrder({ payment: { gatewayId: 'the gateway id', methodId: 'the method id' } })
```

This should work for any of the gateways except ones where the integration is set up to host its own form (Stripe and Ayden, maybe Square as well).

Note that coming into contact with the credit card data does not mean "you are not PCI-compliant". How you interact with data changes the requirements you need to meet to be "PCI-compliant". Technically you need to fill out the SAQ and AOC no matter what when you are a merchant or offering a service that comes into contact with cardholder data in any form (even if your system does not come into direct contact, such as when using a hosted iframe form). If you are using hosted forms, you can fill out SAQ "A" and if you are not using hosted forms you need to fill out "A-EP" (unless you are passing data to your server, then  you need to pass "D" which is 40 pages long and not worth your time, so stick to A-EP compliance). If you only require level 2-4 (which you do if you are even asking about PCI-compliance on a GH issue), then this is sufficient to maintain compliance. So the short answer is: You are probably already PCI-compliant using the non-hosted form, but your business isn't protected if you don't fill out and upload the SAQ/AOC.

You should still switch to hosted forms for your implementation as it further reduces your "PCI scope" (less requirements are applicable since more are outsourced to the gateway in A vs A-EP). But you can determine priority based on whether you think you are already PCI-compliant for "A-EP". Really it is about proving that you don't pass the card data to your server or have any holes that would allow something on the page to pass that data somewhere other than the gateway (eg, TLS/SSL for all assets, scanned any other JS on the page, etc). That is why "A" is much easier to prove, because there is no way you can do that with a hosted form, but as long as you are following web security best practices, you are likely A-EP compliant. So really it is a question of whether it is easier for you to correctly fill out "A-EP" or implement the hosted form integration that you should use to determine your priority.

TLDR: The old way BC did the checkout SDK you needed "A-EP". The hosted forms allows you to just be "A". Both are relatively easy to prove, but "A" is much easier. You never needed something difficult like "D".

See the SAQ forms here:
[A](https://www.pcisecuritystandards.org/documents/PCI-DSS-v3_2_1-SAQ-A.pdf)
[A-EP](https://www.pcisecuritystandards.org/documents/PCI-DSS-v3_2_1-SAQ-A_EP.pdf)

All that said, IANAL or PCI security expert. Take everything I said with a grain of salt and check https://www.pcisecuritystandards.org/ for the most correct and up-to-date info.

**Comment by davidchin** [comment]
Thanks @joeldentici for your detailed reply. @CraigRobertWhite please refer to the example in the reply. You can render a hosted credit card fields by passing in `form` option when initialising the payment step for Braintree.

**Comment by davidchin** [comment]
I forgot to mention, for Braintree, you should pass `form` option under `braintree` key, i.e.: https://github.com/bigcommerce/checkout-sdk-js/blob/master/docs/interfaces/braintreepaymentinitializeoptions.md
